recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,234 @@
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (map[string]string, error)
+}
+
+// LoadSecrets function in SecretsLoader interface will populate resolved secrets based on the Ids returned from FindSecretIDs(). The result is passed as BaseOptions.Secrets to the Driver in Execute() call.
+type SecretsLoader interface {
+	LoadSecrets(map[string]string, ...)(map[string]map[string]string, error){ // returns secretStoreId[key]value
+}
+
+// When calling driver.Execute we pass in BaseOptions which is structured as:
+type BaseOptions struct {
+	// Configuration is the configuration for the recipe.
+	Configuration recipes.Configuration
+
+	// Recipe is the recipe metadata.
+	Recipe recipes.ResourceMetadata
+
+	// Definition is the environment definition for the recipe.
+	Definition recipes.EnvironmentDefinition
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets map[string]map[string]string
+}
+```
+
+### Error Handling
+
+If there is an error when iterating through the recipe configuration or calling ListSecrets API, error will be returned. These errors will not include sensitive data on secret keys or values.
+
+## Test plan
+
+Testing will include unit tests on all new functions created and a
+functional Test will be created to test this functionality e2e.\
+Unit Tests should include test for 'secret not found'
+and verify that upon returning an error, we do not return sensitive information in the error message.
+
+## Security
+
+Secret data will be held in memory and be passed to the Driver which will inject these values into the Terraform configuration in working directory for Terraform. This directory is created when a Terraform recipe is set to be deployed and deleted once deployment of the Terraform recipe is completed.
+
+Secret data will not be persisted and secret data will not be logged by Radius.
+Terraform recipe authors should, in general, set the sensitive flag for any input and output variables that contain sensitive data.
+
+## Compatibility (optional)
+N/A
+
+## Monitoring and Logging
+
+No updates to existing metrics and logs. 
+We need to make sure we do not log sensitive information within calls.
+
+## Development plan
+Updates to Engine, Driver function call - (under 0.5 Sprint)
+Unit Tests + Functional Test - (under 0.5 Sprint)
+
+Total: 1 Sprint
+
+## Open Questions
+
+TBD
+
+## Alternatives considered
+We could also update FindSecretIds() to return a list of SecretStoreIds and the LoadSecrets function retrieves all secrets for the SecretStore and passes the data to the driver.
+We decided to limit the secret data sent to the Driver to only what the Driver will need and are going with the current design. This can be updated in the future if needed.
+
+
+## Design Review Notes
+1. Update design per discussion.Design can be more generic and need not follow Env ConfigStructure.
```

Comment:
Can you call out specific design decisions that were made?

Reply:
updated with some more detail. let me know if the level of detail is okay. Previous versions of the document are also available for users to see.